src/libmongoc/src/mongoc/mongoc-client-pool.c

```diff
@@ -262,7 +266,16 @@ mongoc_client_pool_pop (mongoc_client_pool_t *pool)
          _initialize_new_client (pool, client);
          pool->size++;
       } else {
-         mongoc_cond_wait (&pool->cond, &pool->mutex);
+         if (wait_queue_timeout_ms > 0) {
+            r = mongoc_cond_timedwait (
```

Comment:
I interpreted the `waitQueueTimeoutMS` from CMAP to be the maximum total time a thread should wait to obtain a connection before failing (in our case, we're obtaining a client).

I think we need to track the total time that has passed. It's possible for `mongoc_cond_timedwait` to be called multiple times in the process of trying to obtain a client. For example:
- Thread T1 is waiting for a client, and is sleeping from the call to `mongoc_cond_timedwait`
- Thread T2 pushes a client on, signaling the condition variable.
- Before T1 wakes up, thread T3 pops the client.
- Thread T1 again fails to obtain a client, and sleeps again, but for another period of `maxQueueTimeoutMS`.

One way to do this would be have an `expire_at_ms` at the top.

`expire_at_ms = (bson_get_monotonic_time () / 1000) + wait_queue_timeout_ms;`

Then, before sleeping, check if the current time is after the `expire_at_ms`. If not, only sleep for `expire_at_ms - (current time in MS)`.

Reply:
Excellent point.  I will update the PR to implement the tracking of total wait time.